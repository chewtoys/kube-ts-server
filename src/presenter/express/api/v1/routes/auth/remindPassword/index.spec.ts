import dotenv from 'dotenv';
import assertOnResponseAndStatus from '../../../../../utils/tests/assertOnResponseAndStatus';
dotenv.config();
import { OK } from 'http-status-codes';
import {
  API_V1,
  AUTH,
  REMIND_PASSWORD,
} from '../../../../../../../constants/routes';
import usersFactory from '../../../../../utils/fakeFactories/users/factory';
import initTests from '../../../../../utils/tests/initTests';
import {
  TEST_INVALID_EMAIL,
  TEST_UUID,
  TEST_VALID_EMAIL,
} from '../../../../../utils/tests/testData';

const emailSubject = 'subject';
const link = 'link';

jest.mock('uuid', () => ({
  v4: jest.fn(() => TEST_UUID),
}));

const REMIND_PASSWORD_URL = `${API_V1}${AUTH}${REMIND_PASSWORD}`;

describe('@presenter/auth/register', () => {
  const { request, mailServer, service } = initTests({ useMailServer: true });

  afterAll(() => {
    jest.clearAllMocks();
  });

  it('fails to remind user password without email', async () => {
    await assertOnResponseAndStatus({
      request,
      url: REMIND_PASSWORD_URL,
    });
  });

  it('fails to remind user password when payload contains invalid email', async () => {
    await assertOnResponseAndStatus({
      fields: {
        email: TEST_INVALID_EMAIL,
      },
      request,
      url: REMIND_PASSWORD_URL,
    });

    expect(mailServer.messages.length).toBe(0);
  });

  it('fails to send reset password link when user does not exist', async () => {
    await assertOnResponseAndStatus({
      fields: {
        email: TEST_VALID_EMAIL,
      },
      request,
      statusCode: OK, // @note: for security reasons
      url: REMIND_PASSWORD_URL,
    });
  });

  it('sends reset password link and create reset password token', async () => {
    jest.mock('../../../../../../../translator/factory.ts', () => () => ({
      verifyYourEmailHtml: jest.fn(() => link),
      verifyYourEmailSubject: () => emailSubject,
      verifyYourEmailText: jest.fn(() => link),
    }));

    await usersFactory({
      overrides: {
        email: TEST_VALID_EMAIL,
      },
      service: service.users,
    });


    await assertOnResponseAndStatus({
      fields: {
        email: TEST_VALID_EMAIL,
        id: TEST_UUID
      },
      request,
      statusCode: OK, // @note: for security reasons
      url: REMIND_PASSWORD_URL,
    });

    const { items } = await service.resetPasswordTokens.getItems({
      filter: {
        userId: TEST_UUID
      }
    });

    expect(items.length).toBe(1);
    expect(items[0].id).toBe(TEST_UUID);

    const messages = mailServer.messages;

    expect(messages.length).toBe(1);

    const { textContent, subject, to, from, htmlContent } = messages[0];

    expect(from).toMatchSnapshot('from');
    expect(htmlContent).toMatchSnapshot('htmlContent');
    expect(textContent).toMatchSnapshot('textContent');
    expect(subject).toMatchSnapshot('subject');
    expect(to).toMatchSnapshot('to');
  });
  // tslint:disable-next-line:max-file-line-count
});
