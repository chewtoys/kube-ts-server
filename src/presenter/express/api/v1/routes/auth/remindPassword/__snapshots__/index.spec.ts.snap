// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@presenter/auth/remindPassword fails to remind user password when payload contains invalid email 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "email",
      "message": "Invalid email",
    },
  ],
  "message": "Validation failed",
}
`;

exports[`@presenter/auth/remindPassword fails to remind user password without email 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "email",
      "message": "Expected string between 0 and 255 characters long",
    },
  ],
  "message": "Validation failed",
}
`;

exports[`@presenter/auth/remindPassword fails to send reset password link when user does not exist 1`] = `
Object {
  "message": "If the email you specified exists in our system, we've sent a password reset link to it.",
}
`;

exports[`@presenter/auth/remindPassword sends reset password link and create reset password token 1`] = `
Object {
  "message": "If the email you specified exists in our system, we've sent a password reset link to it.",
}
`;

exports[`@presenter/auth/remindPassword sends reset password link and create reset password token: from 1`] = `"noreply@gamechanger.diet"`;

exports[`@presenter/auth/remindPassword sends reset password link and create reset password token: htmlContent 1`] = `"<h1>Hello,</h1><p>There was recently a request to change the password for your account. </p><p>If you requested this password change, please click on the following link to reset your password: <a href=\\"http://localhost:9000/reset-password?token=25742910-c654-4481-949f-c3f01e4e823f&email=valid.email@test.com\\">http://localhost:9000/reset-password?token=25742910-c654-4481-949f-c3f01e4e823f&email=valid.email@test.com</a></p><p>If clicking the link does not work, please copy and paste the URL into your browser instead. </p><p>If you did not make this request, you can ignore this message and your password will remain the same.</p><p>kube-ts-server</p>"`;

exports[`@presenter/auth/remindPassword sends reset password link and create reset password token: subject 1`] = `"[kube-ts-server] Change your password"`;

exports[`@presenter/auth/remindPassword sends reset password link and create reset password token: textContent 1`] = `"Hello, There was recently a request to change the password for your account. If you requested this password change, please click on the following link to reset your password: http://localhost:9000/reset-password?token=25742910-c654-4481-949f-c3f01e4e823f&email=valid.email@test.com. If clicking the link does not work, please copy and paste the URL into your browser instead. If you did not make this request, you can ignore this message and your password will remain the same. kube-ts-server"`;

exports[`@presenter/auth/remindPassword sends reset password link and create reset password token: to 1`] = `"valid.email@test.com"`;
